Added joint_limits_interface - Direct port from ROS 1

ahcorde opened the pull request:
As discussed yesterday in the ros2_control WG, I'm adding this package originally ported by @ddengster

I still need to copy from @ddengster repository the `transmmision_interface` package.

@Karsten1987 @bmagyar 

Co-authored-by: ddengster <ed.fan@osrfoundation.org>

Signed-off-by: ahcorde <ahcorde@gmail.com>

bmagyar commented:
@v-lopez @jordan-palacios could you guys confirm whether the license on this could also be updated?

v-lopez commented:
Yes, no problem.

v-lopez reviewed (changes requested):
I'd cleanup the test part before merging. Also made some minor suggestions.

ahcorde commented:
Sorry for the bad conditions of the package. I was just copying a version from other repo that it was just functional.

I added 
 - common ROS 2 linters
 - Fixed test

destogl commented on joint_limits_interface/CMakeLists.txt:
```diff
@@ -0,0 +1,57 @@
+    rclcpp
+  )
+  add_launch_test(test/joint_limits_rosparam.launch.py)
+  install(TARGETS
```
Could you edit CMakeLists.txt to be somewhat in style with `hardware_interface/CMakeLists.txt`? This could increase readability.

destogl commented on joint_limits_interface/CMakeLists.txt:
```diff
@@ -0,0 +1,57 @@
+    joint_limits_rosparam_test
+    DESTINATION lib/${PROJECT_NAME}
+  )
+  install(FILES
```
as above

destogl commented on joint_limits_interface/CMakeLists.txt:
```diff
@@ -0,0 +1,57 @@
+endif()
+
+# Install headers
+install(DIRECTORY include/${PROJECT_NAME}/
```
as above

destogl commented on joint_limits_interface/CMakeLists.txt:
```diff
@@ -0,0 +1,57 @@
+
+  ament_add_gtest(joint_limits_urdf_test test/joint_limits_urdf_test.cpp)
+  target_include_directories(joint_limits_urdf_test PUBLIC include)
+  ament_target_dependencies(joint_limits_urdf_test
```
as above

destogl commented on joint_limits_interface/README.md:
```diff
@@ -0,0 +1,30 @@
+   - **Loading from URDF** There are convenience methods for loading joint limits information
+     (position, velocity and effort), as well as soft joint limits information from the URDF.
+   - **Loading from ROS params** There are convenience methods for loading joint limits from the ROS parameter server
+     (position, velocity, acceleration, jerk and effort).Parameter specification is the same used in MoveIt,
```
add space between the sentences

destogl commented on joint_limits_interface/include/joint_limits_interface/joint_limits_interface.hpp:
```diff
@@ -0,0 +1,665 @@
+  double max_vel_limit_;
+};
+
+#if 0
```
It seems this code is never used. Do we still need it?

ahcorde replied:
added a TODO

destogl reviewed (commented):
First part of the review.

destogl commented on joint_limits_interface/package.xml:
```diff
@@ -0,0 +1,39 @@
+<package format="2">
```
Why not use `format=3`

ahcorde replied:
The other packages are version 2, I can update this one.

destogl commented on joint_limits_interface/package.xml:
```diff
@@ -0,0 +1,39 @@
+
+  <maintainer email="bence.magyar.robotics@gmail.com">Bence Magyar</maintainer>
+  <maintainer email="enrique.fernandez.perdomo@gmail.com">Enrique Fernandez</maintainer>
+  <maintainer email="mathias.luedtke@ipa.fraunhofer.de">Mathias Lüdtke</maintainer>
```
Add not-active maintainers as authors

ahcorde replied:
can you suggest me which are the authors and maintainer of this package?

destogl replied:
I think Enrique and Mathias are not active so much here, so I would move them in authors. @bmagyar is clearly maintainer :)

destogl commented on joint_limits_interface/test/joint_limits_interface_test.cpp:
```diff
@@ -0,0 +1,638 @@
+
+class JointLimitsHandleTest : public JointLimitsTest, public ::testing::Test {};
+
+// #ifndef NDEBUG // NOTE: This test validates assertion triggering, hence only gets compiled in
```
why not to remove this?

destogl commented on joint_limits_interface/test/joint_limits_interface_test.cpp:
```diff
@@ -0,0 +1,638 @@
+  hardware_interface::JointCommandHandle cmd_handle2;
+  hardware_interface::JointStateHandle state_handle2;
+};
+//
```
remove?

destogl reviewed (changes requested):
Also remove `.pyc` file.

I have to do it one more time

destogl commented:
This looks good to me. I believe using the new concept from #121 this code is only temporarily. If not so there are few more issues. I list them here just for the "documentation":

* headers alpha sort
* There are some `TODO` comments in the code
* use of new JointHandle from #134

destogl commented on joint_limits_interface/test/joint_limits_interface_test.cpp:
```diff
@@ -0,0 +1,615 @@
+// Copyright 2013, PAL Robotics S.L. All rights reserved.
+// All rights reserved.
+//
+// Software License Agreement (BSD License 2.0)
```
Is this potential issue in the `ros2_control` repository? Currently, we have everything Apache2.0 licensed. If so, can we discuss this with PAL to change the license?

v-lopez replied:
If it is a problem we have no objections to change it to Apache2.0

destogl replied:
@bmagyar

destogl reviewed (approved):
It looks good!

guru-florida commented:
Unfortunately it looks like this merge didn't make it in time before the use of the new JointHandle from #134 took place (see also #155). I will see if I can resolve.

destogl commented:
@ahcorde can we close this PR an open the new one wir changes from @guru-florida?


@guru-florida the changes in #134 are also kind of temporary since we will move everything to components in the future (when ResourceManager is finished - #164 and #147). Nevertheless, this restructuring you did is anyway needed since the logic behind is similar.

Maybe you have some idea how to integrate the limits into components. The example implementation of the PositionJoint is in #140

guru-florida commented:
Thanks @destogl. I have created a new [PR181](https://github.com/ros-controls/ros2_control/pull/181).

bmagyar commented:
Closing in favour of #181
